File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    private int[][] directions = new int[][]{{1,0}, {-1,0}, {0,1}, {0,-1}};
    private int m, n;

    private int dfs(char[][] grid, int row, int col){
        if(row < 0 || row == m || col < 0 || col == n || grid[row][col] == '0'){
            return 0;
        }
        int sizeOfIsland = 1;
        grid[row][col] = '0';

        for(int[] direction : directions){
            int newRow = row + direction[0], newCol = col + direction[1];
            sizeOfIsland += dfs(grid, newRow, newCol);
        }
        return sizeOfIsland;
    }

    public int numIslands(char[][] grid) {
        int numberOfIslands = 0;
        char[][] newGrid = Arrays.stream(grid).map(char[]::clone).toArray(char[][]::new);
        m = grid.length;
        n = grid[0].length;

        for(int i = 0; i < m; ++i){
            for(int j = 0; j < n; ++j){
                if(newGrid[i][j] == '1'){
                    dfs(newGrid, i, j);
                    numberOfIslands++;
                }
            }
        }

        return numberOfIslands;
    }
}

File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        m, n = len(heights), len(heights[0])
        directions = [(0,1), (0,-1), (1,0), (-1,0)]
        pacific, atlantic = set(), set()

        def dfs(row: int, col: int, ocean_set: set) -> None:
            ocean_set.add((row, col))

            for dr, dc in directions:
                new_row, new_col = row + dr, col + dc

                if(0 <= new_row < m and 0 <= new_col < n and
                    (new_row, new_col) not in ocean_set and
                    heights[new_row][new_col] >= heights[row][col]):

                    dfs(new_row, new_col, ocean_set)

        for i in range(m):
            dfs(i, 0, pacific)
            dfs(i, n-1, atlantic)
        
        for j in range(n):
            dfs(0, j, pacific)
            dfs(m-1, j, atlantic)
        
        return [list(cell) for cell in (pacific & atlantic)]
        